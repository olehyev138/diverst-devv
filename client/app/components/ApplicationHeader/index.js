/*
 * LoginPage
 *
 * This is the first thing users see of our App, at the "/login" route
 */

import React from "react";
import { connect } from "react-redux";
import { compose } from "redux";
import { createStructuredSelector } from "reselect";
import AccountCircle from "@material-ui/icons/AccountCircle";
import MenuItem from "@material-ui/core/MenuItem";
import Menu from "@material-ui/core/Menu";
import AppBar from "@material-ui/core/AppBar";
import Toolbar from "@material-ui/core/Toolbar";
import IconButton from "@material-ui/core/IconButton";
import Typography from "@material-ui/core/Typography";
import { withStyles } from "@material-ui/core/styles";
import { fade } from "@material-ui/core/styles/colorManipulator";
import { push } from "connected-react-router";
import { handleLogOut, setCurrentUser } from "containers/App/actions";
import Logo from "components/Logo";
import Grid from "@material-ui/core/Grid";

import {
    makeSelectToken,
    makeSelectUser,
    makeSelectEnterprise
}
from "containers/Users/HomePage/selectors";

import defaultLogo from "images/diverst.png";
import styled from 'styled-components';
const drawerWidth = 240;
const styles = theme => ({
    root: {
        width: "100%"
    },
    grow: {
        flexGrow: 1
    },
    menuButton: {
        marginLeft: -12,
        marginRight: 20
    },
    title: {
        fontSize: 14,
        display: "none",
        [theme.breakpoints.up("sm")]: {
            display: "block"
        }
    },
    search: {
        position: "relative",
        borderRadius: theme.shape.borderRadius,
        backgroundColor: fade(theme.palette.common.white, 0.15),
        "&:hover": {
            backgroundColor: fade(theme.palette.common.white, 0.25)
        },
        marginRight: theme.spacing.unit * 2,
        marginLeft: 0,
        width: "100%",
        [theme.breakpoints.up("sm")]: {
            marginLeft: theme.spacing.unit * 3,
            width: "auto"
        }
    },
    searchIcon: {
        width: theme.spacing.unit * 9,
        height: "100%",
        position: "absolute",
        pointerEvents: "none",
        display: "flex",
        alignItems: "center",
        justifyContent: "center"
    },
    inputRoot: {
        color: "inherit",
        width: "100%"
    },
    inputInput: {
        paddingTop: theme.spacing.unit,
        paddingRight: theme.spacing.unit,
        paddingBottom: theme.spacing.unit,
        paddingLeft: theme.spacing.unit * 10,
        transition: theme.transitions.create("width"),
        width: "100%",
        [theme.breakpoints.up("md")]: {
            width: 200
        }
    },
    sectionDesktop: {
        display: "none",
        [theme.breakpoints.up("md")]: {
            display: "flex"
        }
    },
    sectionMobile: {
        display: "flex",
        [theme.breakpoints.up("md")]: {
            display: "none"
        }
    },
    card: {
        minWidth: 275
    },
    bullet: {
        display: "inline-block",
        margin: "0 2px",
        transform: "scale(0.8)"
    },
    pos: {
        marginBottom: 12
    },
    appBar: {
        zIndex: theme.zIndex.drawer + 1,
    },
    drawer: {
        width: drawerWidth,
        flexShrink: 0,
    },
    drawerPaper: {
        width: drawerWidth,
    },
    content: {
        flexGrow: 1,
        padding: theme.spacing.unit * 3,
    },
    toolbar: theme.mixins.toolbar,
    nested: {
        paddingLeft: theme.spacing.unit * 4,
    },
});

export class ApplicationHeader extends React.PureComponent {

    constructor(props) {
        super(props);
        this.state = {
            anchorEl: null
        };
        this.handleLogOut = this.handleLogOut.bind(this);
        this.handleVisitAdmin = this.handleVisitAdmin.bind(this);
        this.handleVisitHome = this.handleVisitHome.bind(this);
    }

    handleLogOut() {
        this.props.handleLogOut(this.props.currentUser);
    }

    handleVisitAdmin() {
        this.props.handleVisitAdmin();
    }

    handleVisitHome() {
        this.props.handleVisitHome();
    }

    handleProfileMenuOpen = event => {
        this.setState({ anchorEl: event.currentTarget });
    };

    handleMenuClose = () => {
        this.setState({ anchorEl: null });
        this.handleMobileMenuClose();
    };

    handleMobileMenuOpen = event => {
        this.setState({ mobileMoreAnchorEl: event.currentTarget });
    };

    handleMobileMenuClose = () => {
        this.setState({ mobileMoreAnchorEl: null });
    };

    render() {
        const { anchorEl } = this.state;
        const { classes, enterprise, showAdmin, position } = this.props;
        const isMenuOpen = Boolean(anchorEl);

        const renderMenu = (
            <Menu
                anchorEl={anchorEl}
                anchorOrigin={{ vertical: "top", horizontal: "right" }}
                transformOrigin={{ vertical: "top", horizontal: "right" }}
                open={isMenuOpen}
                onClose={this.handleMenuClose}
            >
                <MenuItem onClick={this.handleMenuClose}>Profile</MenuItem>
                <MenuItem onClick={this.handleLogOut}>Log Out</MenuItem>
            </Menu>
        );

        return (
            <div>
                <AppBar position={position} className={classes.appBar}>
                    <Toolbar>
                        <Logo imgClass="tiny-img" margin={20} />
                        <Typography
                            className={classes.title}
                            variant="h6"
                            color="inherit"
                            noWrap
                        >
                            {enterprise.name}
                        </Typography>
                        <div className={classes.grow} />
                        <div>
                            <Grid container direction="row" alignItems="center">
                            { showAdmin ?
                                    <div>
                                        <Grid item onClick={this.handleVisitHome}>
                                            Home
                                        </Grid>
                                        <Grid item>
                                            <IconButton
                                                aria-owns={
                                                    isMenuOpen
                                                        ? "material-appbar"
                                                        : undefined
                                                }
                                                aria-haspopup="true"
                                                onClick={this.handleProfileMenuOpen}
                                                color="inherit"
                                            >
                                                <AccountCircle />
                                            </IconButton>
                                        </Grid>
                                    </div>
                                    :
                                    
                                    <div>
                                        <Grid item onClick={this.handleVisitAdmin}>
                                            Admin
                                        </Grid>
                                        <Grid item>
                                            <IconButton
                                                aria-owns={
                                                    isMenuOpen
                                                        ? "material-appbar"
                                                        : undefined
                                                }
                                                aria-haspopup="true"
                                                onClick={this.handleProfileMenuOpen}
                                                color="inherit"
                                            >
                                                <AccountCircle />
                                            </IconButton>
                                        </Grid>
                                    </div>
                                }
                            </Grid>
                        </div>
                    </Toolbar>
                </AppBar>
                {renderMenu}
            </div>
        );
    }
}

export function mapDispatchToProps(dispatch, ownProps) {
    return {
        handleLogOut: function(token) {
            dispatch(handleLogOut(token));
            dispatch(setCurrentUser(null));
        },
        handleVisitAdmin: function() {
            dispatch(push("/admins/analytics"));
        },
        handleVisitHome: function() {
            dispatch(push("/users/home"));
        }
    };
}

const mapStateToProps = createStructuredSelector({
    token: makeSelectToken(),
    currentUser: makeSelectUser(),
    enterprise: makeSelectEnterprise()
});

const withConnect = connect(
    mapStateToProps,
    mapDispatchToProps,
);

export default compose(
    withConnect,
)(withStyles(styles)(ApplicationHeader));
