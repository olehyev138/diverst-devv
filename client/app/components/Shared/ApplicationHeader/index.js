import React, { memo } from 'react';
import { connect } from 'react-redux';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { createStructuredSelector } from 'reselect';
import classNames from 'classnames';

import { withStyles } from '@material-ui/core/styles';
import {
  AppBar, Box, Button, Hidden, IconButton, Link, ListItemIcon, Menu, MenuItem, Toolbar, Typography, CardActionArea
} from '@material-ui/core';
import AccountCircleIcon from '@material-ui/icons/AccountCircle';
import PermIdentityIcon from '@material-ui/icons/PermIdentity';
import ExitToAppIcon from '@material-ui/icons/ExitToApp';
import BuildIcon from '@material-ui/icons/Build';
import DvrIcon from '@material-ui/icons/Dvr';
import MenuIcon from '@material-ui/icons/Menu';

import WrappedNavLink from 'components/Shared/WrappedNavLink';

import Logo from 'components/Shared/Logo/index';
import { logoutBegin } from 'containers/Shared/App/actions';

import { selectEnterprise, selectUser, selectPermissions } from 'containers/Shared/App/selectors';

import { selectGroup } from 'containers/Group/selectors';

import { ROUTES } from 'containers/Shared/Routes/constants';

import DiverstFormattedMessage from 'components/Shared/DiverstFormattedMessage';
import messages from 'containers/Shared/App/messages';
import { resolveRootManagePath } from 'utils/adminLinkHelpers';
import Permission from 'components/Shared/DiverstPermission';

const styles = theme => ({
  grow: {
    flexGrow: 1,
  },
  centerText: {
    textAlign: 'center',
  },
  sectionDesktop: {
    display: 'flex',
  },
  appBar: {
    zIndex: theme.zIndex.drawer + 1,
  },
  content: {
    flexGrow: 1,
    padding: theme.spacing(3),
  },
  toolbar: theme.mixins.toolbar,
  paper: {
    borderWidth: 1,
    borderStyle: 'solid',
    borderColor: theme.custom.colors.grey,
  },
  buttonSection: {
    display: 'flex',
    alignItems: 'center',
  },
  dashboardSwitchButton: {
    color: '#FFFFFF',
    borderColor: '#FFFFFF',
    height: '40px',
    padding: '0px 8px',
    marginRight: '8px',
  },
  dashboardSwitchMenuItem: {
    [theme.breakpoints.up('sm')]: {
      display: 'none',
    },
  },
  dashboardIcon: {
    marginRight: 4,
    fontSize: 20,
  },
  adminIcon: {
    marginRight: 2,
    fontSize: 18,
  },
  drawerToggle: {
    marginRight: theme.spacing(2),
    [theme.breakpoints.up('md')]: {
      display: 'none',
    },
  },
  navLinkActive: {
    backgroundColor: 'rgba(0, 0, 0, 0.14)',
  },
  whiteButton: {
    color: '#ffffff',
  }
});

export class ApplicationHeader extends React.PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      drawerOpen: props.drawerOpen,
      menuAnchor: null,
    };

    this.logoutBegin = this.logoutBegin.bind(this);
  }

  logoutBegin() {
    this.props.logoutBegin();
  }

  handleProfileMenuOpen = (event) => {
    this.setState({ menuAnchor: event.currentTarget });
  };

  handleProfileMenuClose = () => {
    this.setState({ menuAnchor: null });
  };

  handleDrawerToggle = () => {
    this.setState(
      state => ({ drawerOpen: !state.drawerOpen }),
      () => (this.props.drawerToggleCallback(this.state.drawerOpen))
    );
  };

  render() {
    const { menuAnchor } = this.state;
    const {
      classes, enterprise, group, position, isAdmin, user, permissions
    } = this.props;
    const isMenuOpen = Boolean(menuAnchor);

    const adminPath = resolveRootManagePath(permissions);

    const renderMenu = (
      <Menu
        classes={{
          paper: classes.paper,
        }}
        id='userMenu'
        disableAutoFocusItem
        anchorEl={menuAnchor}
        getContentAnchorEl={null}
        elevation={0}
        anchorOrigin={{ vertical: 'bottom', horizontal: 'center' }}
        transformOrigin={{ vertical: 'top', horizontal: 'center' }}
        open={isMenuOpen}
        onClose={this.handleProfileMenuClose}
      >
        <MenuItem
          component={WrappedNavLink}
          to={isAdmin ? ROUTES.user.root.path() : ROUTES.admin.root.path()}
          className={classes.dashboardSwitchMenuItem}
        >
          { isAdmin
            ? (
              <React.Fragment>
                <ListItemIcon>
                  <DvrIcon />
                </ListItemIcon>
                {<DiverstFormattedMessage {...messages.header.dashboard} />}
              </React.Fragment>
            )
            : (
              <React.Fragment>
                <ListItemIcon>
                  <BuildIcon />
                </ListItemIcon>
                {<DiverstFormattedMessage {...messages.header.admin} />}
              </React.Fragment>
            )
          }
        </MenuItem>
        <MenuItem
          component={WrappedNavLink}
          to={ROUTES.user.show.path(user.user_id)}
          activeClassName={classes.navLinkActive}
        >
          <ListItemIcon>
            <PermIdentityIcon />
          </ListItemIcon>
          {<DiverstFormattedMessage {...messages.header.profile} />}
        </MenuItem>
        <MenuItem onClick={this.logoutBegin}>
          <ListItemIcon>
            <ExitToAppIcon />
          </ListItemIcon>
          {<DiverstFormattedMessage {...messages.header.logout} />}
        </MenuItem>
      </Menu>
    );

    return (
      <React.Fragment>
        <AppBar position={position} className={classes.appBar}>
          <Toolbar className={classes.toolbar}>
            { isAdmin
              ? (
                <IconButton
                  aria-label='Open drawer'
                  edge='start'
                  onClick={this.handleDrawerToggle}
                  className={classNames(classes.drawerToggle, classes.whiteButton)}
                >
                  <MenuIcon />
                </IconButton>
              )
              : <React.Fragment />
            }
            <Logo height='55px' withLink />
            <div className={classNames(classes.grow, classes.centerText)}>
              <Hidden xsDown>
                {!isAdmin && group ? (
                  <Typography variant='h5'>
                    {group.name}
                  </Typography>
                )
                  : (<React.Fragment />)
                }
              </Hidden>
            </div>
            <div className={classes.sectionDesktop}>
              <div className={classes.buttonSection}>
                <Permission show={isAdmin || !!adminPath}>
                  <Hidden xsDown>
                    <Link
                      component={WrappedNavLink}
                      to={isAdmin ? ROUTES.user.root.path() : adminPath && adminPath}
                    >
                      <Button
                        className={classes.dashboardSwitchButton}
                        variant='outlined'
                      >
                        { isAdmin
                          ? (
                            <span>
                              <DvrIcon className={classes.dashboardIcon} />
                              {<DiverstFormattedMessage {...messages.header.dashboard} />}
                            </span>
                          )
                          : (
                            <span>
                              <BuildIcon className={classes.adminIcon} />
                              {<DiverstFormattedMessage {...messages.header.admin} />}
                            </span>
                          )
                        }
                      </Button>
                    </Link>
                  </Hidden>
                </Permission>
                <div>
                  <IconButton
                    aria-controls={
                      isMenuOpen
                        ? 'userMenu'
                        : undefined
                    }
                    aria-haspopup='true'
                    onClick={this.handleProfileMenuOpen}
                    className={classes.whiteButton}
                  >
                    <AccountCircleIcon />
                  </IconButton>
                </div>
              </div>
            </div>
          </Toolbar>
        </AppBar>
        {renderMenu}
      </React.Fragment>
    );
  }
}

ApplicationHeader.propTypes = {
  classes: PropTypes.object,
  user: PropTypes.object,
  group: PropTypes.object,
  permissions: PropTypes.object,
  drawerOpen: PropTypes.bool,
  drawerToggleCallback: PropTypes.func,
  enterprise: PropTypes.object,
  position: PropTypes.string,
  isAdmin: PropTypes.bool,
  logoutBegin: PropTypes.func,
  handleVisitAdmin: PropTypes.func,
  handleVisitHome: PropTypes.func,
};

ApplicationHeader.defaultProps = {
  position: 'absolute'
};

export function mapDispatchToProps(dispatch, ownProps) {
  return {
    logoutBegin() {
      dispatch(logoutBegin());
    },
  };
}

const mapStateToProps = createStructuredSelector({
  user: selectUser(),
  enterprise: selectEnterprise(),
  group: selectGroup(),
  permissions: selectPermissions(),
});

const withConnect = connect(
  mapStateToProps,
  mapDispatchToProps,
);

export const StyledApplicationHeader = withStyles(styles)(ApplicationHeader);

export default compose(
  withConnect,
  withStyles(styles),
  memo,
)(ApplicationHeader);
