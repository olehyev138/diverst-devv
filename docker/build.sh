#!/bin/bash

# Set path from second argument or default to current path
BUILD_PATH=${2:-.}

if [ -n "$1" ]; then
  docker build -t "$1" "$BUILD_PATH" "${@:3}"
else
  echo "ERROR: Requires a docker image tag as first argument (such as diverstdev:X.X)" 1>&2
  exit 1
fi
