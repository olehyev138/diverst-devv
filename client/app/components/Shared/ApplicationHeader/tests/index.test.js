import React from 'react';
import { shallow } from 'enzyme/build';
import { unwrap } from '@material-ui/core/test-utils/index';

import { StyledApplicationHeader } from 'components/Shared/ApplicationHeader/index';
const ApplicationHeaderNaked = unwrap(StyledApplicationHeader);

/**
 * TODO:
 *   - test button clicks dispatch correct actions
 */

const props = {
  classes: {},
  user: {},
  drawerOpen: true,
  drawerToggleCallback: jest.fn(),
  enterprise: {},
  position: 'position',
  isAdmin: true,
  logoutBegin: jest.fn(),
  handleVisitAdmin: jest.fn(),
  handleVisitHome: jest.fn()
};

describe('<ApplicationHeader />', () => {
  it('should not log errors in console', () => {
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = shallow(<ApplicationHeaderNaked {...props} />);

    expect(spy).not.toHaveBeenCalled();
  });

  it('should render and match the snapshot', () => {
    const wrapper = shallow(<ApplicationHeaderNaked {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
