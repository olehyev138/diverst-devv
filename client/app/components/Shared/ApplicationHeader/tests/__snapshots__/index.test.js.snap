// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ApplicationHeader /> should render and match the snapshot 1`] = `
<div>
  <WithStyles(ForwardRef(AppBar))
    position="position"
  >
    <WithStyles(ForwardRef(Toolbar))>
      <WithStyles(ForwardRef(IconButton))
        aria-label="Open drawer"
        color="inherit"
        edge="start"
        onClick={[Function]}
      >
        <MenuIcon />
      </WithStyles(ForwardRef(IconButton))>
      <WithStyles(ForwardRef(Link))
        component={
          Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          }
        }
        to="/"
      >
        <WithStyles(ForwardRef(Button))>
          <Connect(WithStyles(Logo))
            imgClass="large-img"
            verticalPadding={20}
          />
        </WithStyles(ForwardRef(Button))>
      </WithStyles(ForwardRef(Link))>
      <div
        className=""
      >
        <Hidden
          xsDown={true}
        />
      </div>
      <div>
        <div>
          <Hidden
            xsDown={true}
          >
            <WithStyles(ForwardRef(Link))
              component={
                Object {
                  "$$typeof": Symbol(react.forward_ref),
                  "render": [Function],
                }
              }
              to="/"
            >
              <WithStyles(ForwardRef(Button))
                variant="outlined"
              >
                <span>
                  <DvrIcon />
                  Dashboard
                </span>
              </WithStyles(ForwardRef(Button))>
            </WithStyles(ForwardRef(Link))>
          </Hidden>
          <div>
            <WithStyles(ForwardRef(IconButton))
              aria-haspopup="true"
              color="inherit"
              onClick={[Function]}
            >
              <AccountCircleIcon />
            </WithStyles(ForwardRef(IconButton))>
          </div>
        </div>
      </div>
    </WithStyles(ForwardRef(Toolbar))>
  </WithStyles(ForwardRef(AppBar))>
  <WithStyles(ForwardRef(Menu))
    anchorEl={null}
    anchorOrigin={
      Object {
        "horizontal": "center",
        "vertical": "bottom",
      }
    }
    classes={
      Object {
        "paper": undefined,
      }
    }
    disableAutoFocusItem={true}
    elevation={0}
    getContentAnchorEl={null}
    id="userMenu"
    onClose={[Function]}
    open={false}
    transformOrigin={
      Object {
        "horizontal": "center",
        "vertical": "top",
      }
    }
  >
    <WithStyles(ForwardRef(MenuItem))
      component={
        Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        }
      }
      to="/"
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <DvrIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      Dashboard
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <PermIdentityIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      Profile
    </WithStyles(ForwardRef(MenuItem))>
    <WithStyles(ForwardRef(MenuItem))
      onClick={[Function]}
    >
      <WithStyles(ForwardRef(ListItemIcon))>
        <ExitToAppIcon />
      </WithStyles(ForwardRef(ListItemIcon))>
      Log Out
    </WithStyles(ForwardRef(MenuItem))>
  </WithStyles(ForwardRef(Menu))>
</div>
`;
